compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -1690,93 +1863,85 @@ private STNode validateForUsageOfVar(STNode typeDesc) {
-                // Fall through
             case OBJECT_KEYWORD:
-            case ABSTRACT_KEYWORD: // Allow `abstract` to be parsed as an object type qualifier and then log an error
-            case CLIENT_KEYWORD:
-                return parseObjectTypeDescriptor();
+                return parseObjectTypeDescriptor(qualifiers);
             case OPEN_PAREN_TOKEN:
+                reportInvalidQualifierList(qualifiers);
                 return parseNilOrParenthesisedTypeDesc();
             case MAP_KEYWORD: // map type desc
             case FUTURE_KEYWORD: // future type desc
+                reportInvalidQualifierList(qualifiers);
                 return parseParameterizedTypeDescriptor();
-            case TYPEDESC_KEYWORD: // typedesc type desc
+            case TYPEDESC_KEYWORD:
+                reportInvalidQualifierList(qualifiers);
                 return parseTypedescTypeDescriptor();
-            case ERROR_KEYWORD: // error type descriptor
+            case ERROR_KEYWORD:
+                reportInvalidQualifierList(qualifiers);
                 return parseErrorTypeDescriptor();
-            case XML_KEYWORD: // typedesc type desc
+            case XML_KEYWORD:
+                reportInvalidQualifierList(qualifiers);
                 return parseXmlTypeDescriptor();
-            case STREAM_KEYWORD: // stream type desc
+            case STREAM_KEYWORD:
+                reportInvalidQualifierList(qualifiers);
                 return parseStreamTypeDescriptor();
-            case TABLE_KEYWORD: // table type desc
+            case TABLE_KEYWORD:
+                reportInvalidQualifierList(qualifiers);
                 return parseTableTypeDescriptor();
             case FUNCTION_KEYWORD:
-            case TRANSACTIONAL_KEYWORD:
-                return parseFunctionTypeDesc();
+                return parseFunctionTypeDesc(qualifiers);
             case OPEN_BRACKET_TOKEN:
+                reportInvalidQualifierList(qualifiers);
                 return parseTupleTypeDesc();
             case DISTINCT_KEYWORD:
-                return parseDistinctTypeDesc(context);
-            case ISOLATED_KEYWORD:
-                if (isFuncDefOrFuncTypeStart()) {
-                    return parseFunctionTypeDesc();
-                } else {
-                    return parseObjectTypeDescriptor();
-                }
+                reportInvalidQualifierList(qualifiers);
+                STNode distinctKeyword = consume();
+                return parseDistinctTypeDesc(distinctKeyword, context);
             default:
+                if (isTypeDescQualifier(nextToken.kind)) {
+                    // Type desc qualifiers are present. Parse them first and then try to re-parse the method.
+                    parseTypeDescQualifiers(qualifiers);
+                    return parseTypeDescriptorInternal(qualifiers, context, isInConditionalExpr);
+                }
```

Comment:
What if we move this, entirely before the switch-case, and only call `parseTypeDescQualifiers(qualifiers);`, (without the if-check and without the re-parse)?
inside `parseTypeDescQualifiers(qualifiers)` we can have a while loop, rather than the do-while.

Reply:
Can do the same for top-level nodes as well